### ISSUE TITLE
Add GAIT date fetching to test plan.

### ISSUE THREAD

**Issue by sanjuroj** [mercycorps/TolaActivity#1418]
We would like the current behavior from the Django admin to be replicated in the new PA admin.  There are several key features of the current behavior.
- When a program is saved, the GAIT ID entered by the user is used to query the GAIT database and the program start and end dates are extracted from that information
- The dates are used to populate both the "GAIT program dates" and "Program start and end dates" areas of the Program period modal. Note that GAIT program dates and Program start and end dates can be different since GAIT program dates can include mid-month values.
- If dates are already present in those fields, they are not overwritten by the dates just retrieved from GAIT.
- An error message is provided if the GAIT dates could not be retrieved for some reason.
- I'm assuming that since the dates are not displayed in the new program admin directly, the decision was made to have them only appear in the reporting period modal.

**Comment by RickieT** [comment]
1. If a valid GAIT ID is entered (2 that I stole from other programs after deleting them) and saved, when I return to the program page or country page in which the program was created, the GAIT program dates say "Unavailable". Neither program or gait start/end dates were populated

2. Regardless of what is entered as a "GAIT ID", there is no messaging displayed, only displays and error if the ID is a duplicate when trying to save. 

3. If I enter a GAIT ID that is invalid (1233124) in the Django Admin, an error is displayed telling me that it could not fetch GAIT Information for that ID, this does not occur in Admin Lite.

**Comment by jennyhavah** [comment]
@sanjuroj Hoping you can take over testing this one, as you know the most about how all of this is supposed to work.

**Comment by sanjuroj** [comment]
I don't think any of the GAIT-date-fetching code was ported over to the new program admin.

**Comment by KennyMonster** [comment]
In the new admin, what does the messaging look like for success/failure cases of getting dates from the GAIT server? I'm assuming we want to continue to allow the creation of new programs even if the GAIT server is unreachable or does not return dates for a given ID?

**Comment by sanjuroj** [comment]
Yes, programs should still be created, even if the GAIT fetch fails.  I'm not sure that the existing admin ever actually surfaced any error messages, but that might be easier in the new admin.  But it would be an enhancement of the existing functionality.

**Comment by KennyMonster** [comment]
The existing Django admin did surface an error, but did not report success.

![Screen Shot 2019-04-22 at 4 36 24 PM](https://user-images.githubusercontent.com/1278224/56540220-ed864480-651c-11e9-96fb-a5648484e6a1.png)

The new admin has a focus on form validation with regards to error handling, or throwing errors into the PNotify, which is also used for creation success. As such, I'm not sure where to throw a GAIT error message exactly. Options are the top of the form, or a secondary PNotify message.

**Comment by sanjuroj** [comment]
I don't think that's been something that's been talked about yet.  I'm sort of partial to the PNotify errors, they seem more visible to me.

**Comment by jennyhavah** [comment]
Not sure how we handled this in the past, but we need to keep mind that some programs don't have GAIT ID's for legitimate reasons, and that doesn't warrant a PNotify error message. The way that gets surfaced to end users is with the "Unavailable" message in the program period modal. Not sure what was happening in the Django admin side.

Now, if there's some sort of connection error or other technical reason why the system fails to complete the check, then _that_ seems like it merits a PNotify error message -- ideally with some sort of option like "Try again".

**Comment by KennyMonster** [comment]
I'm thinking we make an API endpoint for pulling GAIT dates that runs independent of the Program creation endpoint, which is currently how it's set up now. The API endpoint in turn could be run automatically on program creation, but could also be run via a "Pull GAIT dates" button in the UI. I don't think it makes sense to run automatically during Program update.

Having a dedicated endpoint would also make it much easier to return specific status to the uses such as "success", "unavailable", or "GAIT server/network error".

This would only be implemented for the new admin, as I don't think it makes sense to add any functionality to the Django admin code path.

**Comment by RickieT** [comment]
@RickieT

**Comment by KennyMonster** [comment]
@jennyhavah So because this started out as a test plan ticket, I've been implementing stuff without input from design, so I wanted to check in where things are at.

As implemented:

* Existing functionality in the Django Admin remains
* The GAIT date syncing functionality has been added to the new admin on program creation
* There are handful of paths and subsequent UI feedback to the admin creating a program
  * A user does not enter in a GAIT id into the form - GAIT date syncing is avoided and the form works as it currently does
  * A user enters in a GAIT ID and the dates are successfully synced over
  * A user enters in a GAIT ID and a failure happens during the sync process

Below are screen shots for each of the 3 bullet points when a program is saved

No GAIT ID:
![Screen Shot 2019-04-24 at 11 24 50 AM](https://user-images.githubusercontent.com/1278224/56684265-3d335000-6684-11e9-90c3-6bc440127010.png)

Sync success:
![Screen Shot 2019-04-24 at 11 24 33 AM](https://user-images.githubusercontent.com/1278224/56684280-47ede500-6684-11e9-9398-059fcd648782.png)

Sync failure:
![Screen Shot 2019-04-24 at 11 23 50 AM](https://user-images.githubusercontent.com/1278224/56684291-50462000-6684-11e9-817e-742292c7d370.png)

In the last screen shot, the string after the ":" is variable - the reason being is there are various reasons why a sync may fail such as

* The request to Tola fails because network/server
* The request to GAIT server fails because network/server
* The GAIT ID doesn't exist on the GAIT server (invalid ID)

I thought it may be useful to report which case happened so when the user contacts the Tola team for support, we have a little bit of an idea where to start helping them.

Open questions:

* Is the verbiage for the feedback ok? Do we even want this level of feedback?
* This was implemented in such a way that a "retry sync" button is relatively easy to add at this point, I just need to know the placement and verbiage, or if it is a post-ship nice-to-have feature.

**Comment by jennyhavah** [comment]
Looking good, @KennyMonster! I'm with you on trying to provide error messages that are as useful as possible. My suggestions on wording are below.

# No GAIT ID

Successfully saved  (sentence case)

# Sync success

Successfully synced GAIT program start and end dates

# Sync failure

Successfully saved   (sentence case)

Failed to sync GAIT program start and end dates: There was a network or server connection error.   

Failed to sync GAIT program start and end dates: There was a problem connecting to the GAIT server.

Failed to sync GAIT program start and end dates: The GAIT ID could not be found.

**Comment by jennyhavah** [comment]
@KennyMonster If you can knock it out quickly, I think a "Retry sync" option is a great idea. I would include it inside the PNotify itself, as a simple text link or button.

**Comment by jennyhavah** [comment]
A note from Marie that may or may not be relevant to this ticket. Tagging @RickieT and @KennyMonster 

> ...when I created a new program, nothing happened - no screen closure, no success message, nothing to indicate to me that system successfully saved my program. So, I tried saving again and it told me that there was already a program with my GAIT ID. I went ahead and saved anyway. I saw the "Success Message" that time. I checked the program list, and only 1 of my saves stuck. I expected to see 2 programs but only one appeared, which I guess is good. But, this makes me wonder what happened the first time I tried to save.

**Comment by jennyhavah** [comment]
Actually, this may be an issue of Marie getting used to the new PNotify success messages. Moving this back into the testing queue.

**Comment by RickieT** [comment]
**QA - nearly complete**
* Waiting on some server development to verify GAIT ID successfuly pulls gait dates, but our other options are in verified working
* Duplicate GAIT ID warning:
![image](https://user-images.githubusercontent.com/23220180/56695344-aa9faa80-669d-11e9-8b98-9f1597ae0a7c.png)
* Saved, but failed to sync from GAIT
![image](https://user-images.githubusercontent.com/23220180/56695355-b4291280-669d-11e9-8359-1868acc83204.png)
